panther_battery/src/adc_node.py

```diff
@@ -186,8 +205,42 @@ def _publish_battery_msg(
         else:
             battery_msg.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_DISCHARGING
 
+        # check battery health
+        with self._lock:
+            error_msg = None
+            if V_bat_mean < self.V_BAT_FATAL_MIN:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_DEAD
+                error_msg = 'Battery voltage is critically low!'
+            elif V_bat_mean > self.V_BAT_FATAL_MAX:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_OVERVOLTAGE
+                error_msg = 'Battery overvoltage!'
+            elif temp_bat >= self._high_bat_temp:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_OVERHEAT
+                error_msg = 'Battery temperature is dangerously high!'
+            elif self._driver_battery_last_info_time is None:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_UNKNOWN
+            elif (
+                rospy.get_time() - self._driver_battery_last_info_time < 0.1
+                and abs(V_bat_mean - self._V_driv_mean) > self.V_BAT_FATAL_MAX * 0.1
+            ):
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_UNSPEC_FAILURE
+            else:
+                battery_msg.power_supply_health = BatteryState.POWER_SUPPLY_HEALTH_GOOD
+
+        if error_msg is not None:
+            rospy.logerr_throttle_identical(10.0, f'[{rospy.get_name()}] {error_msg}')
+
         bat_pub.publish(battery_msg)
 
+    def _count_volt_mean(self, label: Union[rospy.Publisher, str], new_val: float) -> float:
+        self._V_bat_mean[label] += (
+            -self._V_bat_hist[label][0] / self._volt_mean_length + new_val / self._volt_mean_length
```

Comment:
A comment would be nice. I had trouble understanding what is going on. Anyway, cool approach to the problem